component/mcsd/updater.go

```diff
@@ -48,31 +50,80 @@ func buildUpdateTransaction(ctx context.Context, tx *fhir.Bundle, entries []fhir
 		}
 		log.Ctx(ctx).Debug().Msgf("Adding entry #%d to transaction: %s", i, resourceType)
 
-		setResourceMetaSource(resource, "")
+		updateResourceMeta(resource, *entry.FullUrl)
 		// Get or create local reference
-		// TODO: Lookup ID local to local mCSD Directory, not remote
-		//localResourceID := localRefMap[*entry.FullUrl]
-		//if localResourceID == "" {
-		//	localResourceID = fmt.Sprintf("urn:uuid:%s", uuid.NewString())
-		//	localRefMap[*entry.FullUrl] = localResourceID
-		//}
-		//resource["id"] = localResourceID
+		// TODO: If the resource already exists, we should look up the existing resource's ID
+		// TODO: We should scope resource IDs to the source (e.g. by prefixing with the source URL or a hash thereof), because when syncing from multiple sources, IDs may collide.
+		remoteLocalRef := resourceType + "/" + resource["id"].(string)
+		localResourceID := remoteRefToLocalRefMap[remoteLocalRef]
+		if localResourceID == "" {
+			localResourceID = generateLocalID()
+			remoteRefToLocalRefMap[remoteLocalRef] = localResourceID
+		}
+		resource["id"] = localResourceID
+		if err := normalizeReferences(resource, remoteRefToLocalRefMap); err != nil {
+			return nil, fmt.Errorf("failed to normalize references in entry #%d: %w", i, err)
+		}
+
 		resourceJSON, _ := json.Marshal(resource)
```

Comment:
We probably do not want to swallow marshalling errors?

Reply:
we unmarshalled it, so it can be marshalled again, but returning the err won't hurt here